File: docker/web/webrtc_video/src/send.py
import asyncio
import json
import websockets
import gi

gi.require_version('Gst', '1.0')
gi.require_version('GstWebRTC', '1.0')
from gi.repository import Gst, GstWebRTC, GstSdp

class WebRTCSend:
    def __init__(self):
        self.SIGNALING_SERVER = 'ws://0.0.0.0:8765'
        self.websocket = None
        self.webrtcbin = None
        self.pipeline = None

    async def connect(self):
        """Establish WebSocket connection to the signaling server."""
        try:
            self.websocket = await websockets.connect(self.SIGNALING_SERVER)
            print("Connected to signaling server")
        except Exception as e:
            print(f"Failed to connect to signaling server: {e}")

        #await self.websocket.send(json.dumps('HELLO'))

    def on_negotiation_needed(self, _):
        """Triggered when WebRTC negotiation is needed."""
        print("Negotiation needed, creating offer...")
        promise = Gst.Promise.new_with_change_func(self.on_offer_created, self.webrtcbin, None)
        self.webrtcbin.emit("create-offer", None, promise)

    def on_offer_created(self, promise, _, __):
        """Handles SDP offer creation."""
        print("Offer created, setting local description...")
        promise.wait()
        reply = promise.get_reply()

        # Because of older GStreamer version on NVIDIA, using second variant
        #offer = reply['offer']
        offer = reply.get_value('offer')

        # Set the local description
        promise = Gst.Promise.new()
        self.webrtcbin.emit('set-local-description', offer, promise)
        promise.interrupt()

        # Send the SDP offer to the remote peer
        self.send_sdp_offer(offer)

    def send_sdp_offer(self, offer):
        """Send the SDP offer to the remote peer over WebSocket."""
        if not self.websocket:
            print("No WebSocket connection, cannot send offer.")
            return
        
        text = offer.sdp.as_text()
        print ('Sending offer:\n%s' % text)
        msg = json.dumps({'sdp': {'type': 'offer', 'sdp': text}})
        try:
            loop = asyncio.new_event_loop()
            loop.run_until_complete(self.websocket.send(msg))
            loop.close()
        except Exception as e:
            print(f"Failed to send SDP offer: {e}")

    def start_pipeline(self):
        """Sets up and starts the GStreamer pipeline with WebRTC."""
        # Create the GStreamer pipeline
        self.pipeline = Gst.parse_launch('webrtcbin name=sendrecv bundle-policy=max-bundle latency=0 \
            stun-server=stun://stun.l.google.com:19302 \
            v4l2src device=/dev/cam-arducam ! video/x-raw,width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! nvv4l2h264enc bitrate=2300000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            \
            v4l2src device=/dev/cam-microdia ! video/x-raw,width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! nvv4l2h264enc bitrate=2300000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            \
            nvarguscamerasrc sensor-mode=4 ! video/x-raw(memory:NVMM),width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! nvv4l2h264enc bitrate=2300000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            \
            v4l2src device=/dev/video20 ! video/x-raw,width=640,height=360,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! nvv4l2h264enc bitrate=2300000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv.')

        # Working pipeline for CPU encoding (x264enc)

        # self.pipeline = Gst.parse_launch('webrtcbin name=sendrecv bundle-policy=max-bundle latency=0 \
        #     stun-server=stun://stun.l.google.com:19302 \
        #     v4l2src device=/dev/video0 ! video/x-raw,width=640,height=480,framerate=30/1 \
        #     ! videoconvert ! video/x-raw,format=I420 ! queue max-size-buffers=1 max-size-time=20000000 max-size-bytes=0 leaky=downstream \
        #     ! x264enc tune=zerolatency speed-preset=ultrafast rc-lookahead=0 bitrate=1000 key-int-max=30 qp-min=18 qp-max=25 \
        #     ! h264parse ! rtph264pay config-interval=1 pt=96 \
        #     ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv.')
        
        # self.pipeline = Gst.parse_launch('webrtcbin name=sendrecv bundle-policy=max-bundle stun-server=stun://stun.l.google.com:19302 videotestsrc is-live=true ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay ! queue ! application/x-rtp,media=video,encoding-name=VP8,payload=97 ! sendrecv. \
        # videotestsrc is-live=true ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay ! queue ! application/x-rtp,media=video,encoding-name=VP8,payload=97 ! sendrecv. \
        # audiotestsrc is-live=true ! audioconvert ! audioresample ! opusenc ! rtpopuspay ! application/x-rtp,media=audio,encoding-name=OPUS,payload=96 ! sendrecv.')
        # Get the webrtcbin element
        
        self.webrtcbin = self.pipeline.get_by_name('sendrecv')

        # Can't use latency property on NVIDIA because of old GStreamer version
        #self.webrtcbin.set_property("latency", 0)
        self.webrtcbin.set_property("bundle-policy", "max-bundle")
        self.webrtcbin.set_property("stun-server", "stun://stun.l.google.com:19302")


        # Connect to signals
        self.webrtcbin.connect('on-negotiation-needed', self.on_negotiation_needed)
        self.webrtcbin.connect('on-ice-candidate', self.send_ice_candidate)
        self.webrtcbin.connect('pad-added', self.on_incoming_stream)

        # Start the pipeline (this is where the negotiation would kick off)
        self.pipeline.set_state(Gst.State.PLAYING)

    def handle_sdp(self, msg):
        """Handle incoming SDP answer."""
        sdp = msg['sdp']
        assert(sdp['type'] == 'answer')
        sdp = sdp['sdp']
        print ('Received answer:\n%s' % sdp)

        # Process the answer into a WebRTCSessionDescription
        res, sdpmsg = GstSdp.SDPMessage.new()
        GstSdp.sdp_message_parse_buffer(bytes(sdp.encode()), sdpmsg)
        answer = GstWebRTC.WebRTCSessionDescription.new(GstWebRTC.WebRTCSDPType.ANSWER, sdpmsg)

        # Set the remote description (the answer)
        promise = Gst.Promise.new()
        self.webrtcbin.emit('set-remote-description', answer, promise)
        promise.interrupt()
    
    def handle_ice(self, msg):
        ice = msg['ice']
        candidate = ice['candidate']
        sdpmlineindex = ice['sdpMLineIndex']
        self.webrtcbin.emit('add-ice-candidate', sdpmlineindex, candidate)

    def send_ice_candidate(self, _, mlineindex, candidate):
        """Send ICE candidate to the remote peer over WebSocket."""
        print(f"Sending ICE candidate: {candidate}")
        icemsg = json.dumps({'ice': {'candidate': candidate, 'sdpMLineIndex': mlineindex}})
        loop = asyncio.new_event_loop()
        loop.run_until_complete(self.websocket.send(icemsg))
        loop.close()

    def on_incoming_stream(self, _, pad):
        """Handle incoming stream from the remote peer."""
        if pad.direction != Gst.PadDirection.SRC:
            return

        decodebin = Gst.ElementFactory.make('decodebin')
        decodebin.connect('pad-added', self.on_incoming_decodebin_stream)
        self.pipeline.add(decodebin)
        decodebin.sync_state_with_parent()
        self.webrtcbin.link(decodebin)

    def on_incoming_decodebin_stream(self, _, pad):
        """Handle incoming decodebin stream."""
        if not pad.has_current_caps():
            print(pad, 'has no caps, ignoring')
            return

        caps = pad.get_current_caps()
        assert len(caps)
        s = caps[0]
        name = s.get_name()

        if name.startswith('video'):
            pass
            # q = Gst.ElementFactory.make('queue')
            # conv = Gst.ElementFactory.make('videoconvert')
            # sink = Gst.ElementFactory.make('autovideosink')
            
            # # Minimize latency in queue (reduce buffering)
            # q.set_property("max-size-buffers", 1)
            # q.set_property("max-size-time", 0)
            # q.set_property("max-size-bytes", 0)
            # q.set_property("leaky", "downstream")  # Allow data to drop if too much buffering happens
            
            # # Disable sync on autovideosink for lower latency
            # sink.set_property("sync", False)

            # self.pipeline.add(q, conv, sink)
            # self.pipeline.sync_children_states()
            # pad.link(q.get_static_pad('sink'))
            # q.link(conv)
            # conv.link(sink)

        elif name.startswith('audio'):
            pass
            # q = Gst.ElementFactory.make('queue')
            # conv = Gst.ElementFactory.make('audioconvert')
            # resample = Gst.ElementFactory.make('audioresample')
            # sink = Gst.ElementFactory.make('autoaudiosink')
            
            # # Same low-latency settings for audio queue
            # q.set_property("max-size-buffers", 1)
            # q.set_property("max-size-time", 0)
            # q.set_property("max-size-bytes", 0)
            # q.set_property("leaky", "downstream")

            # self.pipeline.add(q, conv, resample, sink)
            # self.pipeline.sync_children_states()
            # pad.link(q.get_static_pad('sink'))
            # q.link(conv)
            # conv.link(resample)
            # resample.link(sink)
    
    async def listen(self):
        """Main loop to handle incoming messages."""
        async for message in self.websocket:
            msg = json.loads(message)
            if msg == {"status": "HELLO"}:
                print("Received HELLO from RECV peer")
                await self.websocket.send(json.dumps({'status': 'OK'}))
                self.start_pipeline()
            elif 'sdp' in msg:
                self.handle_sdp(msg)
            elif 'ice' in msg:
                self.handle_ice(msg)
            elif msg.startswith('ERROR'):
                print(msg)
                return 1
        return 0

if __name__ == "__main__":
    Gst.init(None)
    webrtc_send = WebRTCSend()
    loop = asyncio.get_event_loop()
    loop.run_until_complete(webrtc_send.connect())
    loop.run_until_complete(webrtc_send.listen())

File: docker/web/webrtc_video/src/send.cpp
// Required headers
#include <iostream>
#include <gst/gst.h>
#include <gst/webrtc/webrtc.h>
#include <gst/sdp/gstsdp.h>
#include <jsoncpp/json/json.h>
#include <websocketpp/config/asio_no_tls_client.hpp>
#include <websocketpp/client.hpp>
#include <X11/Xlib.h>
#include <glib.h>
#include <thread>
#include <atomic>
#include <mutex>

/**
 * @brief Main class handling WebRTC video streaming
 * 
 * This class manages the WebRTC connection, GStreamer pipeline,
 * and WebSocket signaling to stream video from multiple cameras
 */
class WebRTCSend {
public:
    /**
     * @brief Constructor - initializes GStreamer and member variables
     */
    WebRTCSend() {
        // Initialize GStreamer
        gst_init(nullptr, nullptr);
        this->pipeline = nullptr;
        this->webrtcbin = nullptr;
        ws_running = true;
    }

    /**
     * @brief Destructor - cleans up resources
     */
    ~WebRTCSend() {
        ws_running = false;
        ws_cv.notify_all();
        if (ws_thread.joinable()) {
            ws_thread.join();
        }
        if (pipeline) {
            gst_element_set_state(pipeline, GST_STATE_NULL);
            gst_object_unref(pipeline);
        }
        if (webrtcbin) {
            gst_object_unref(webrtcbin);
        }
    }

    /**
     * @brief Establishes WebSocket connection to signaling server
     */
    void connect() {
        try {
            // Set up WebSocket connection using websocketpp
            websocketpp::client<websocketpp::config::asio_client> client;
            client.init_asio();

            client.set_open_handler(std::bind(&WebRTCSend::on_open, this, std::placeholders::_1, &client));
            client.set_message_handler(std::bind(&WebRTCSend::on_msg, this, std::placeholders::_1, std::placeholders::_2));

            websocketpp::lib::error_code ec;
            websocketpp::client<websocketpp::config::asio_client>::connection_ptr con = client.get_connection("ws://0.0.0.0:8765", ec);

            if (ec) {
                std::cout << "Connection error: " << ec.message() << std::endl;
            }

            client.connect(con);
            client.run();

        } catch (const std::exception& e) {
            std::cout << "Failed to connect: " << e.what() << std::endl;
        }
    }

    /**
     * @brief Handler called when WebSocket connection is established
     * 
     * @param hdl Connection handle
     * @param c Pointer to WebSocket client
     */
    void on_open(websocketpp::connection_hdl hdl, websocketpp::client<websocketpp::config::asio_client>* c) {
        std::cout << "Connected! Waiting for HELLO message..." << std::endl;

        global_hdl = hdl;
        global_client = c;

        ws_thread = std::thread(&WebRTCSend::process_ws_queue, this);
    }

    /**
     * @brief Handler for incoming WebSocket messages
     * 
     * Processes different types of messages including:
     * - HELLO handshake
     * - ICE candidates
     * - SDP offers/answers
     */
    void on_msg(websocketpp::connection_hdl, websocketpp::client<websocketpp::config::asio_client>::message_ptr msg) {
        std::string payload = msg->get_payload();

        Json::CharReaderBuilder reader;
        Json::Value jsonMsg;
        std::string errs;
        std::istringstream s(payload);

        if (Json::parseFromStream(reader, s, &jsonMsg, &errs)) {
            if (!jsonMsg.isObject()) {
                std::cout << "Error: JSON message is not an object" << std::endl;
                return;
            }

            if (jsonMsg.isMember("status") && jsonMsg["status"].asString() == "HELLO") {
                std::cout << "Received HELLO." << std::endl;
                Json::Value reply;
                reply["status"] = "OK";
                Json::StreamWriterBuilder writer;
                std::string message = Json::writeString(writer, reply);
                queue_ws(message);

                // Post pipeline setup to main thread
                g_idle_add([](gpointer data) -> gboolean {
                    static_cast<WebRTCSend*>(data)->setup_pipeline();
                    return G_SOURCE_REMOVE;
                }, this);

            } else if (jsonMsg.isMember("ice")) {
                std::cout << "Received ICE candidate." << std::endl;
                // Copy payload for lambda capture
                std::string payload_copy = payload;
                g_idle_add([](gpointer data) -> gboolean {
                    auto* pair = static_cast<std::pair<WebRTCSend*, std::string>*>(data);
                    pair->first->handle_ice(pair->second);
                    delete pair;
                    return G_SOURCE_REMOVE;
                }, new std::pair<WebRTCSend*, std::string>(this, payload_copy));

            } else if (jsonMsg.isMember("sdp")) {
                std::cout << "Received SDP answer." << std::endl;
                std::string payload_copy = payload;
                g_idle_add([](gpointer data) -> gboolean {
                    auto* pair = static_cast<std::pair<WebRTCSend*, std::string>*>(data);
                    pair->first->handle_sdp(pair->second);
                    delete pair;
                    return G_SOURCE_REMOVE;
                }, new std::pair<WebRTCSend*, std::string>(this, payload_copy));

            } else {
                std::cout << "Unknown JSON message type" << std::endl;
            }
        } else {
            std::cout << "Failed to parse JSON: " << errs << std::endl;
        }
    }

    /**
     * @brief Processes queued WebSocket messages
     * 
     * Runs in a separate thread to handle outgoing WebSocket messages
     */
    void process_ws_queue() {
        while (ws_running) {
            std::unique_lock<std::mutex> lock(ws_mutex);
            ws_cv.wait(lock, [this] { return !msg_queue.empty() || !ws_running; });

            std::string msg = msg_queue.front();
            msg_queue.pop();
            lock.unlock();

            websocketpp::lib::error_code ec;
            global_client->send(global_hdl, msg, websocketpp::frame::opcode::text, ec);
            if (ec) {
                std::cerr << "Error sending WebSocket message: " << ec.message() << std::endl;
            } else {
                std::cout << "Sent message over WebSocket: " << msg << std::endl;
            }
        }
    }

    // Commented out direct send method in favor of queued approach
    // void send_ws(const std::string& msg) {
    //     websocketpp::lib::error_code ec;
    //     global_client->send(global_hdl, msg, websocketpp::frame::opcode::text, ec);
    //     if (ec) {
    //         std::cerr << "Error sending WebSocket message: " << ec.message() << std::endl;
    //     } else {
    //         std::cout << "Sent message over WebSocket: " << msg << std::endl;
    //     }
    // }

    /**
     * @brief Queues a message to be sent over WebSocket
     * 
     * @param msg Message to be queued
     */
    void queue_ws(const std::string& msg) {
        std::lock_guard<std::mutex> lock(ws_mutex);
        msg_queue.push(msg);
        ws_cv.notify_one();
    }

    /**
     * @brief Sets up the GStreamer pipeline for video streaming
     * 
     * Creates and configures a pipeline with:
     * - Multiple v4l2src sources for different cameras
     * - NVIDIA video conversion and H264 encoding
     * - WebRTC transmission
     */
    void setup_pipeline() {
        // Create GStreamer pipeline
        GError* error = nullptr;
        pipeline = gst_parse_launch("webrtcbin name=sendrecv bundle-policy=max-bundle latency=0 \
            stun-server=stun://stun.l.google.com:19302 \
            v4l2src device=/dev/cam-arducam ! video/x-raw,width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! queue max-size-buffers=2 leaky=downstream \
            ! nvv4l2h264enc bitrate=2500000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! queue max-size-buffers=2 leaky=downstream \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            \
            nvarguscamerasrc sensor-mode=4 ! video/x-raw(memory:NVMM),width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! queue max-size-buffers=2 leaky=downstream \
            ! nvv4l2h264enc bitrate=2500000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! queue max-size-buffers=2 leaky=downstream \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            \
            v4l2src device=/dev/cam-aveo ! video/x-raw,width=640,height=480,framerate=30/1 \
            ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            ! queue max-size-buffers=2 leaky=downstream \
            ! nvv4l2h264enc bitrate=2500000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            ! queue max-size-buffers=2 leaky=downstream \
            ! h264parse ! rtph264pay config-interval=1 pt=96 \
            ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv.",
            &error);

            // v4l2src device=/dev/cam-arducam ! video/x-raw,width=640,height=480,framerate=30/1 \
            // ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            // ! nvv4l2h264enc bitrate=3000000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            // ! h264parse ! rtph264pay config-interval=1 pt=96 \
            // ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            // \
            // v4l2src device=/dev/cam-microdia ! video/x-raw,width=640,height=480,framerate=30/1 \
            // ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            // ! nvv4l2h264enc bitrate=3000000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            // ! h264parse ! rtph264pay config-interval=1 pt=96 \
            // ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv. \
            // \
            // nvarguscamerasrc sensor-mode=4 ! video/x-raw(memory:NVMM),width=640,height=480,framerate=30/1 \
            // ! nvvidconv ! video/x-raw(memory:NVMM),format=NV12 \
            // ! nvv4l2h264enc bitrate=3000000 iframeinterval=30 control-rate=1 preset-level=1 profile=2 maxperf-enable=true \
            // ! h264parse ! rtph264pay config-interval=1 pt=96 \
            // ! application/x-rtp,media=video,encoding-name=H264,payload=96 ! sendrecv.

            // nvcompositor name=mix sync-import-streams=false \
            // sink_0::xpos=0   sink_0::ypos=0   sink_0::width=640  sink_0::height=480 \
            // sink_1::xpos=640 sink_1::ypos=0   sink_1::width=640  sink_1::height=480 \
            // sink_2::xpos=1280 sink_2::ypos=0  sink_2::width=640  sink_2::height=480 \
            // ! queue max-size-buffers=7 leaky=upstream \
            // ! video/x-raw(memory:NVMM), width=1920, height=480, framerate=30/1 \
            // ! nvvidconv \
            // ! video/x-raw(memory:NVMM), format=NV12 \
            // ! queue max-size-buffers=7 leaky=upstream \
            // ! nvv4l2h264enc maxperf-enable=true bitrate=4000000 idrinterval=5 iframeinterval=5 insert-sps-pps=true insert-aud=true insert-vui=true \
            // ! queue max-size-buffers=7 leaky=upstream \
            // ! h264parse config-interval=1 \
            // ! rtph264pay pt=96 mtu=1200 config-interval=1 \
            // ! sendrecv. \
            // \
            // nvarguscamerasrc sensor-mode=4 \
            // ! queue max-size-buffers=7 leaky=upstream \
            // ! video/x-raw(memory:NVMM),width=640,height=480,framerate=30/1 \
            // ! nvvidconv \
            // ! video/x-raw(memory:NVMM),format=NV12 \
            // ! mix.sink_0 \
            // \
            // v4l2src device=/dev/video9 io-mode=4 \
            // ! video/x-raw,width=640,height=480,framerate=30/1 \
            // ! nvvidconv \
            // ! video/x-raw(memory:NVMM),format=NV12 \
            // ! mix.sink_1 \
            // \
            // v4l2src device=/dev/video7 io-mode=4 \
            // ! video/x-raw,width=640,height=480,framerate=30/1 \
            // ! nvvidconv \
            // ! video/x-raw(memory:NVMM),format=NV12 \
            // ! mix.sink_2", &error);

        if (error) {
            std::cerr << "ERROR: Could not create GStreamer pipeline: " << error->message << std::endl;
            g_error_free(error);
            return;
        }

        if (!pipeline) {
            std::cerr << "ERROR: Could not create GStreamer pipeline." << std::endl;
            return;
        }

        // Get webrtcbin element
        webrtcbin = gst_bin_get_by_name(GST_BIN(pipeline), "sendrecv");

        if (!webrtcbin) {
            std::cerr << "ERROR: Could not get WebRTC element." << std::endl;
            return;
        }

        gst_pipeline_use_clock(GST_PIPELINE(pipeline), gst_system_clock_obtain());

        // Set WebRTC properties
        g_object_set(G_OBJECT(webrtcbin), "bundle-policy", GST_WEBRTC_BUNDLE_POLICY_MAX_BUNDLE, "stun-server", "stun://stun.l.google.com:19302", nullptr);

        // Connect to signals
        g_signal_connect(webrtcbin, "on-negotiation-needed", G_CALLBACK(&WebRTCSend::on_negotiation_needed), this);
        g_signal_connect(webrtcbin, "on-ice-candidate", G_CALLBACK(&WebRTCSend::send_ice_candidate), this);
        g_signal_connect(webrtcbin, "pad-added", G_CALLBACK(&WebRTCSend::on_incoming_stream), this);

        // Set pipeline state to PLAYING
        gst_element_set_state(pipeline, GST_STATE_PLAYING);
    }

    /**
     * @brief Callback for sending ICE candidates
     * 
     * Called when a new ICE candidate is discovered
     */
    static void send_ice_candidate(GstElement* webrtcbin, guint mlineindex, gchar* candidate, gpointer user_data) {
        std::cout << "Sending ICE candidate: " << candidate << std::endl;

        // Create a JSON message to send over WebSocket
        Json::Value icemsg;
        icemsg["ice"]["candidate"] = candidate;
        icemsg["ice"]["sdpMLineIndex"] = static_cast<int>(mlineindex);  // Casting mlineindex to int for JSON

        // Convert JSON message to string
        Json::StreamWriterBuilder writer;
        std::string icemsg_str = Json::writeString(writer, icemsg);

        // Send the ICE candidate over WebSocket
        static_cast<WebRTCSend*>(user_data)->queue_ws(icemsg_str);
    }

    /**
     * @brief Handles incoming ICE candidates
     * 
     * @param ice JSON string containing ICE candidate information
     */
    void handle_ice(const std::string& ice) {
        Json::CharReaderBuilder reader;
        Json::Value jsonMsg;
        std::string errs;
        std::istringstream s(ice);

        if (Json::parseFromStream(reader, s, &jsonMsg, &errs)) {
            std::string candidate = jsonMsg["ice"]["candidate"].asString();
            int sdpMLineIndex = jsonMsg["ice"]["sdpMLineIndex"].asInt();

            if (!webrtcbin) {
                std::cerr << "ERROR: Could not get WebRTC element." << std::endl;
                return;
            }

            g_signal_emit_by_name(webrtcbin, "add-ice-candidate", sdpMLineIndex, candidate.c_str());
        } else {
            std::cerr << "Failed to parse ICE candidate: " << errs << std::endl;
        }
    }

    /**
     * @brief Callback when WebRTC negotiation is needed
     */
    static void on_negotiation_needed(GstElement* webrtcbin, gpointer user_data) {
        std::cout << "Negotiation needed" << std::endl;

        GstPromise* promise = gst_promise_new_with_change_func(on_offer_created, user_data, nullptr);
        g_signal_emit_by_name(webrtcbin, "create-offer", nullptr, promise);
    }

    /**
     * @brief Callback when SDP offer is created
     */
    static void on_offer_created(GstPromise* promise, gpointer user_data) {
        GstWebRTCSessionDescription* offer = nullptr;
        const GstStructure* reply = gst_promise_get_reply(promise);
        gst_structure_get(reply, "offer", GST_TYPE_WEBRTC_SESSION_DESCRIPTION, &offer, nullptr);
        gst_promise_unref(promise);

        if (!offer) {
            std::cerr << "Failed to create offer" << std::endl;
            return;
        }

        gchar* sdp_str = gst_sdp_message_as_text(offer->sdp);
        std::cout << "Created offer: " << sdp_str << std::endl;

        Json::Value sdpmsg;
        sdpmsg["sdp"]["type"] = "offer";
        sdpmsg["sdp"]["sdp"] = sdp_str;

        Json::StreamWriterBuilder writer;
        std::string sdpmsg_str = Json::writeString(writer, sdpmsg);

        static_cast<WebRTCSend*>(user_data)->queue_ws(sdpmsg_str);

        g_signal_emit_by_name(static_cast<WebRTCSend*>(user_data)->webrtcbin, "set-local-description", offer, nullptr);

        g_free(sdp_str);
        gst_webrtc_session_description_free(offer);
    }

    /**
     * @brief Handles incoming SDP messages
     * 
     * @param sdp JSON string containing SDP information
     */
    void handle_sdp(const std::string& sdp) {
        Json::CharReaderBuilder reader;
        Json::Value jsonMsg;
        std::string errs;
        std::istringstream s(sdp);

        if (Json::parseFromStream(reader, s, &jsonMsg, &errs)) {
            std::string sdp_str = jsonMsg["sdp"]["sdp"].asString();
            GstSDPMessage* sdp_msg = nullptr;
            gst_sdp_message_new(&sdp_msg);
            GstSDPResult result = gst_sdp_message_parse_buffer((guint8*)sdp_str.c_str(), sdp_str.size(), sdp_msg);

            if (result != GST_SDP_OK) {
                std::cerr << "Failed to parse SDP message: " << result << std::endl;
                return;
            }

            GstWebRTCSessionDescription* answer = gst_webrtc_session_description_new(GST_WEBRTC_SDP_TYPE_ANSWER, sdp_msg);
            if (!answer) {
                std::cerr << "Failed to create WebRTC session description" << std::endl;
                gst_sdp_message_free(sdp_msg);
                return;
            }

            g_signal_emit_by_name(webrtcbin, "set-remote-description", answer, nullptr);
            gst_webrtc_session_description_free(answer);
        } else {
            std::cerr << "Failed to parse SDP answer: " << errs << std::endl;
        }
    }

    /**
     * @brief Callback for handling incoming media streams
     */
    static void on_incoming_stream(GstElement* webrtcbin, GstPad* pad, WebRTCSend* self) {
        std::cout << "Received incoming stream." << std::endl;

        GstPad* sinkpad;
        GstElement* decodebin = gst_element_factory_make("decodebin", nullptr);
        gst_bin_add(GST_BIN(self->pipeline), decodebin);
        gst_element_sync_state_with_parent(decodebin);

        g_signal_connect(decodebin, "pad-added", G_CALLBACK(on_decodebin_pad_added), self);

        sinkpad = gst_element_get_static_pad(decodebin, "sink");
        gst_pad_link(pad, sinkpad);
        gst_object_unref(sinkpad);
    }

    /**
     * @brief Callback when decodebin creates a new pad
     * 
     * Sets up appropriate elements for handling decoded audio/video streams
     */
    static void on_decodebin_pad_added(GstElement* decodebin, GstPad* pad, WebRTCSend* self) {
        GstCaps* caps = gst_pad_get_current_caps(pad);
        const GstStructure* str = gst_caps_get_structure(caps, 0);
        const gchar* name = gst_structure_get_name(str);

        GstElement* conv = nullptr;
        GstElement* sink = nullptr;

        if (g_str_has_prefix(name, "video")) {
            conv = gst_element_factory_make("videoconvert", nullptr);
            sink = gst_element_factory_make("autovideosink", nullptr);
        } else if (g_str_has_prefix(name, "audio")) {
            conv = gst_element_factory_make("audioconvert", nullptr);
            sink = gst_element_factory_make("autoaudiosink", nullptr);
        }

        if (conv && sink) {
            gst_bin_add_many(GST_BIN(self->pipeline), conv, sink, nullptr);
            gst_element_sync_state_with_parent(conv);
            gst_element_sync_state_with_parent(sink);

            GstPad* sinkpad = gst_element_get_static_pad(conv, "sink");
            gst_pad_link(pad, sinkpad);
            gst_object_unref(sinkpad);

            GstPad* srcpad = gst_element_get_static_pad(conv, "src");
            GstPad* sinkpad2 = gst_element_get_static_pad(sink, "sink");
            gst_pad_link(srcpad, sinkpad2);
            gst_object_unref(srcpad);
            gst_object_unref(sinkpad2);
        }

        gst_caps_unref(caps);
    }

private:
    GstElement* pipeline;                // Main GStreamer pipeline
    GstElement* webrtcbin;              // WebRTC element
    websocketpp::connection_hdl global_hdl;  // WebSocket connection handle
    websocketpp::client<websocketpp::config::asio_client>* global_client;  // WebSocket client

    std::queue<std::string> msg_queue;   // Queue for outgoing WebSocket messages
    std::mutex ws_mutex;                 // Mutex for thread safety
    std::condition_variable ws_cv;       // Condition variable for message queue
    std::thread ws_thread;               // Thread for processing WebSocket messages
    std::atomic<bool> ws_running{true};  // Flag to control WebSocket thread
};

/**
 * @brief Main entry point
 * 
 * Initializes X11 threading, creates WebRTCSend instance,
 * and runs the main loop
 */
int main() {
    if (!XInitThreads()) {
        std::cerr << "Failed to initialize X11 threading!" << std::endl;
        return 1;
    }

    WebRTCSend recv;
    std::thread ws_thread([&]() { recv.connect(); });

    GMainLoop* loop = g_main_loop_new(nullptr, FALSE);
    g_main_loop_run(loop);
    g_main_loop_unref(loop);

    ws_thread.join();
    return 0;
}